test/unit/sourceset.test.js

```diff
@@ -0,0 +1,532 @@
+import log from '../../src/js/utils/log.js';
+import sinon from 'sinon';
+
+const Html5 = videojs.getTech('Html5');
+const wait = 100;
+let qunitFn = 'module';
+
+if (!Html5.canOverrideAttributes()) {
+  qunitFn = 'skip';
+}
+
+const oldMovingMedia = Html5.prototype.movingMediaElementInDOM;
+const validateSource = function(assert, player, sources, checkMediaElSource = true) {
+  const tech = player.tech_;
+  const mediaEl = tech.el();
+
+  // assert.deepEqual(player.currentSource(), sources[0], 'currentSource is correct');
+  // assert.equal(player.src(), sources[0].src, 'src is correct');
+  // assert.deepEqual(player.currentSources(), sources, 'currentSources is correct');
+  // when we are dealing with <source> elements mediaEl.src will be null
+  if (checkMediaElSource) {
+    assert.equal(mediaEl.src, sources[0].src, 'mediaEl.src is correct');
+    assert.equal(mediaEl.getAttribute('src'), sources[0].src, 'mediaEl attribute is correct');
+    assert.equal(tech.src(), sources[0].src, 'tech is correct');
+  }
+};
+
+const testTypes = ['video el', 'change video el', 'audio el', 'change audio el'];
+
+QUnit[qunitFn]('sourceset', function(hooks) {
+  QUnit.module('source before player', (subhooks) => testTypes.forEach((testName) => {
+    QUnit.module(testName, {
+      beforeEach() {
+        sinon.stub(log, 'error');
+        if (testName === 'change video el' || testName === 'change audio el') {
+          Html5.prototype.movingMediaElementInDOM = false;
+        }
+        this.hook = (player) => player.on('sourceset', () => this.sourcesets++);
+        videojs.hook('setup', this.hook);
+
+        this.sourcesets = 0;
+        this.fixture = document.getElementById('qunit-fixture');
+
+        if ((/audio/i).test(testName)) {
+          this.mediaEl = document.createElement('audio');
+          this.testSrc = {
+            src: 'http://vjs.zencdn.net/v/oceans.mp3',
+            type: 'audio/mpeg'
+          };
+        } else {
+          this.mediaEl = document.createElement('video');
+          this.testSrc = {
+            src: 'http://vjs.zencdn.net/v/oceans.mp4',
+            type: 'video/mp4'
+          };
+        }
+        this.sourceOne = {src: 'http://example.com/one.mp4', type: 'video/mp4'};
+        this.sourceTwo = {src: 'http://example.com/two.mp4', type: 'video/mp4'};
+
+        if ((/audio/).test(testName)) {
+          this.sourceOne = {src: 'http://example.com/one.mp3', type: 'audio/mpeg'};
+          this.sourceTwo = {src: 'http://example.com/two.mp3', type: 'audio/mpeg'};
+        }
+
+        this.mediaEl.className = 'video-js';
+        this.fixture.appendChild(this.mediaEl);
+      },
+      afterEach(assert) {
+        const done = assert.async();
+
+        if (typeof this.totalSourcesets === 'undefined') {
+          this.totalSourcesets = 1;
+        }
+
+        window.setTimeout(() => {
+          assert.equal(this.sourcesets, this.totalSourcesets, 'no additional sourcesets');
```

Comment:
running assertions in the `afterEach` just feels wrong to me.

Reply:
Well we could add logic to each to to make sure we don't get extra source sets, or a function that can be re-used, not sure which route to go.